### PR TITLE
Make unload function of loaded image protocol an Option

### PR DESCRIPTION
**Pull request by joschock**
This change makes a the "unload" function pointer of the loaded image protocol an `Option` in line with the Transpose Guidelines in lib.rs. This pointer is often NULL, which makes it UB to interact with in Rust if it is not an option.

### CODE REVIEW

**Comment by joschock** [comment]
@dvdhrm  - is this an acceptable approach?

**Comment by dvdhrm** [comment]
Thanks for the patch! Generally, it looks good, but 2 issues:

1) The spec does not mention that the callback can be NULL (I couldn't find anything in section 9.1). How does EDK2 handle this? Can all protocol functions be NULL? Certainly not, right? So which ones need to be checked? Am I missing something?

2) This is clearly an API break, so it would require a version bump, unless we consider this a bugfix without which the API cannot be used. Any comments? I do not mind bumping the major version, but maybe we can snuck it in without it.

**Comment by joschock** [comment]
Thanks for looking at it :)

> 1. The spec does not mention that the callback can be NULL (I couldn't find anything in section 9.1). How does EDK2 handle this? Can all protocol functions be NULL? Certainly not, right? So which ones need to be checked? Am I missing something?

That this particular function can be NULL is implied by other verbiage elsewhere in the spec. For example, Section 7.4.3:

> "If the image has been started and has an Unload() entry point, control is passed to that entry point. If the image’s unload function returns EFI_SUCCESS, the image is unloaded; otherwise, the error returned by the image’s unload function
> is returned to the caller. The image unload function is responsible for freeing all allocated memory and ensuring that
> there are no references to any freed memory, or to the image itself, before returning EFI_SUCCESS.
> 
> If the image has been started and **does not have an Unload()** entry point, the function returns EFI_UNSUPPORTED."

The Unload() here is referring to this function pointer in LoadedImage. Essentially the usage model is that drivers that wish to be "unloadable" can signal that by updating the Unload() pointer in LoadedImageProtocol in their EntryPoint, but this is optional. If the image entrypoint has run and returned SUCCESS, and Unload() is still NULL, then the core prevents unloading the image. So NULL in this function pointer serves as a "this image cannot be unloaded once started" marker.

From the EDK2 C reference implementation (https://github.com/tianocore/edk2/blob/dd58d1227c102ad4a799002f16b3a8c796d7d7cc/MdeModulePkg/Core/Dxe/Image/Image.c#L1933), "does not have an Unload() entry point" means "is NULL".

In addition, the SCT is verifying expected behavior with Unload() set to NULL (https://github.com/tianocore/edk2-test/blob/0e2ced3befa431bb1aebff005c4c4f1a9edfe6b4/uefi-sct/SctPkg/TestCase/UEFI/EFI/BootServices/ImageServices/BlackBoxTest/ImageBBTestFunction.c#L341). 

The EDK2 UefiDriverEntryPoint library that nearly all (at least, all EDK2-based C UEFI drivers) use may supply an Unload() (https://github.com/tianocore/edk2/blob/dd58d1227c102ad4a799002f16b3a8c796d7d7cc/MdePkg/Library/UefiDriverEntryPoint/DriverEntryPoint.c#L110), but this again is contingent on whether EDK2 Autogen process thinks the module needs one based on presence of library destructors, UNLOAD_IMAGE in .inf files, etc. - so if you have a very simple driver with none of these elements, you'll end up with a NULL Unload() in your LoadedImage instance corresponding to it.

> 2. This is clearly an API break, so it would require a version bump, unless we consider this a bugfix without which the API cannot be used. Any comments? I do not mind bumping the major version, but maybe we can snuck it in without it.

I agree that it would be an API break (at least on the rust side, the ABI is the same), but I don't have strong opinions about the versioning. I'd be okay with either major bump or not per whatever you think is best for the project.

**Comment by dvdhrm** [comment]
On another note: Do you have contacts at the UEFI forum (or even the editor)? Is there a way to submit errata that might get fixed in an upcoming spec-release? Or is it not worth the effort?

**Comment by joschock** [comment]
>  Do you want to file a PR? Otherwise, I will just pick it up myself and publish a release with a major version bump.

I'm not sure I understand this question - this is a PR already? But I don't have a strong opinion about how the patch goes in, if you would like to cherry-pick it or otherwise bring it in via some other means that is more convenient to the project, I'm totally fine with that.

**Comment by joschock** [comment]
> On another note: Do you have contacts at the UEFI forum (or even the editor)? Is there a way to submit errata that might get fixed in an upcoming spec-release? Or is it not worth the effort?

I don't have direct access, but I chatted about it with one of my colleagues who does and he's planning to submit a change request to clarify the Unload() usage in the spec as per above discussion.

**Comment by joschock** [comment]
@dvdhrm  - I've worked with our team internally to submit a request for change to the spec to clarify that Unload() can be NULL to indicate that the Image does not support Unload(). Are we good to move ahead with this one?

**Comment by dvdhrm** [comment]
Yes, we are. Thanks for pushing this forward.

I somehow mistook this as a github-issue with a referenced commit, sorry for that.

I was a bit hesitant regarding bumping the major version (it will be a mess if other libraries expose r-efi types in their API, and thus you might get issues updating r-efi across multiple dependencies). However, I came to the conclusion that the recommendation from r-efi should be to always avoid exposing r-efi in public APIs, and instead just use some void-pointer (or `*mut ()`, ...). It allows compatibility with non-r-efi libraries and prevents exactly this version-bump issue. Hence, I think bumping the major version is a non-issue.

Long story short: I will merge this and push out a release. Thanks a lot for the report!

**Comment by dvdhrm** [comment]
(merged and pushed; v5.0.0 released as well; thanks a lot!)
